File: app/src/main/java/cl/figonzal/evaluatool/evalua/evalua7/modulo4/velocidadFragments/VelocidadFragmentE7M4.java
/*-----------------------------------------------------------------------------
                                                                              -
 This file is subject to the terms and conditions defined in                  -
 file 'LICENSE', which is part of this source code package                    -
                                                                              -
 Autor: Felipe González                                                       -
 Email: felipe.gonzalezalarcon94@gmail.com                                    -
                                                                              -
 Copyright (c) 2020                                                           -
                                                                              -
 Last modified 14-08-20 1:09                                                  -
 -----------------------------------------------------------------------------*/

package cl.figonzal.evaluatool.evalua.evalua7.modulo4.velocidadFragments;

import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.fragment.app.Fragment;

import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.crashlytics.FirebaseCrashlytics;

import java.util.Objects;

import cl.figonzal.evaluatool.R;
import cl.figonzal.evaluatool.dialogs.CorregidoDialogFragment;
import cl.figonzal.evaluatool.interfaces.EvaluaInterface;
import cl.figonzal.evaluatool.utilidades.Utilidades;

public class VelocidadFragmentE7M4 extends Fragment implements EvaluaInterface {

    private static final double DESVIACION = 68.21;
    private static final double MEDIA = 257.12;
    private final Integer[][] percentil_segundos = new Integer[][]{
            {100, 99},
            {121, 97},
            {135, 95},
            {150, 90},
            {160, 85},
            {170, 80},
            {180, 75},
            {190, 70},
            {210, 65},
            {220, 60},
            {240, 55},
            {250, 50},
            {260, 40},
            {280, 30},
            {300, 20},
            {340, 10},
            {500, 5}
    };
    //TAREA 1
    private TextInputEditText et_segundos_t1;
    private int segundos_t1 = 0;
    //SUBTOTALES
    private TextView tv_sub_total_t1;
    private double total_pd_t1 = 0;
    //TOTALES
    private TextView tv_pd_total;
    private TextView tv_pd_corregido;
    private TextView tv_percentil;
    private TextView tv_nivel;
    private TextView tv_desviacion_calculada;
    private ProgressBar progressBar;

    private FirebaseCrashlytics crashlytics;

    public VelocidadFragmentE7M4() {
        // Required empty public constructor
    }

    static VelocidadFragmentE7M4 newInstance() {
        return new VelocidadFragmentE7M4();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View v = inflater.inflate(R.layout.fragment_velocidad_e7_m4, container, false);

        crashlytics = FirebaseCrashlytics.getInstance();

        instanciarRecursosInterfaz(v);

        textWatcherTarea1();

        return v;
    }

    private void instanciarRecursosInterfaz(View v) {
        TextView tv_media = v.findViewById(R.id.tv_media_value);
        TextView tv_desviacion = v.findViewById(R.id.tv_desviacion_value);
        tv_media.setText(String.valueOf(MEDIA));
        tv_desviacion.setText(String.valueOf(DESVIACION));

        //SUBTOTALES
        tv_sub_total_t1 = v.findViewById(R.id.tv_pd_subtotal_t1);
        et_segundos_t1 = v.findViewById(R.id.et_segundos_t1);

        //TOTALES
        tv_pd_total = v.findViewById(R.id.tv_pd_total_value);
        tv_pd_corregido = v.findViewById(R.id.tv_pd_total_corregido_value);
        tv_percentil = v.findViewById(R.id.tv_percentil_value);
        tv_nivel = v.findViewById(R.id.tv_nivel_obtenido_value);
        tv_desviacion_calculada = v.findViewById(R.id.tv_desviacion_calculada_value);

        progressBar = v.findViewById(R.id.progressBar);
        progressBar.setMax(percentil_segundos[0][1]);

        ImageView iv_corregido = v.findViewById(R.id.iv_help_pd_corregido);
        iv_corregido.setOnClickListener(v1 -> {
            Log.d(getString(R.string.DIALOGO_AYUDA), getString(R.string.DIALOGO_AYUDA_MSG_ABIERTO));

            crashlytics.log(getString(R.string.DIALOGO_AYUDA) + getString(R.string.DIALOGO_AYUDA_MSG_ABIERTO));

            CorregidoDialogFragment dialogFragment = new CorregidoDialogFragment();
            dialogFragment.setCancelable(false);
            dialogFragment.show(Objects.requireNonNull(getActivity()).getSupportFragmentManager(), getString(R.string.DIALOGO_AYUDA));
        });
    }

    private void textWatcherTarea1() {
        et_segundos_t1.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                total_pd_t1 = 0;
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {

                if (s.length() == 0) {
                    segundos_t1 = 0;
                } else if (s.length() > 0) {
                    segundos_t1 = Integer.parseInt(Objects.requireNonNull(et_segundos_t1.getText()).toString());
                }
                total_pd_t1 = calcularTarea(null, tv_sub_total_t1, "Tarea 1: ", segundos_t1, null, null);
                calcularResultado();
            }
        });

    }

    @Override
    public double calcularTarea(Integer n_tarea, TextView tv_sub_total, String tarea, Integer aprobadas, Integer omitidas, Integer reprobadas) {
        tv_sub_total.setText(String.format("%s%s seg", tarea, aprobadas));
        return aprobadas;
    }

    @Override
    public void calcularResultado() {

        double total_pd;
        total_pd = total_pd_t1;

        tv_pd_total.setText(String.format("%s seg", total_pd));


        double pd_corregido = corregirPD(percentil_segundos, total_pd);
        tv_pd_corregido.setText(String.format("%s seg", pd_corregido));

        int percentil = calcularPercentil(pd_corregido);
        tv_percentil.setText(String.valueOf(percentil));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            progressBar.setProgress(percentil, true);
        } else {
            progressBar.setProgress(percentil);
        }
        String nivel = Utilidades.calcularNivel(percentil);
        tv_nivel.setText(nivel);

        double desviacion = Utilidades.calcularDesviacion(MEDIA, DESVIACION, pd_corregido, true);
        tv_desviacion_calculada.setText(String.valueOf(desviacion));
    }

    @Override
    public int calcularPercentil(double pd_total) {
        if (pd_total < percentil_segundos[0][0]) {
            return percentil_segundos[0][1];
        } else if (pd_total > percentil_segundos[percentil_segundos.length - 1][0]) {
            return percentil_segundos[percentil_segundos.length - 1][1];
        } else {
            for (Integer[] item : percentil_segundos) {
                if (pd_total <= item[0]) {
                    return item[1];
                }
            }
        }
        //Percentil no encontrado
        Log.d(getString(R.string.TAG_PERCENTIL_CALCULADO), getString(R.string.PERCENTIL_NULO));

        crashlytics.log(getString(R.string.TAG_PERCENTIL_CALCULADO) + getString(R.string.PERCENTIL_NULO));
        return -1;
    }

    @Override
    public double corregirPD(Integer[][] perc, double pd_actual) {
        if (pd_actual < perc[0][0]) {
            return perc[0][0];
        } else if (pd_actual > perc[perc.length - 1][0]) {
            return perc[perc.length - 1][0];
        } else {
            for (Integer[] item : perc) {
                if (pd_actual <= item[0]) {
                    return item[0];
                }
            }
        }
        Log.d(getString(R.string.TAG_SEG_CORREGIDOS), getString(R.string.SEG_NULOS));

        crashlytics.log(getString(R.string.TAG_SEG_CORREGIDOS) + getString(R.string.SEG_NULOS));
        return -1;
    }
}

File: app/src/main/java/cl/figonzal/evaluatool/evalua/evalua7/modulo4/velocidadFragments/VelocidadFragmentE7M4.kt
/*

 This file is subject to the terms and conditions defined in
 file 'LICENSE', which is part of this source code package

 Autor: Felipe González
 Email: felipe.gonzalezalarcon94@gmail.com

 Copyright (c) 2021

 Last modified 03-05-21 1:46
 */
package cl.figonzal.evaluatool.evalua.evalua7.modulo4.velocidadFragments

import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import androidx.fragment.app.Fragment
import cl.figonzal.evaluatool.R
import cl.figonzal.evaluatool.databinding.FragmentVelocidadE7M4Binding
import cl.figonzal.evaluatool.interfaces.EvaluaInterface
import cl.figonzal.evaluatool.utilidades.Utils
import cl.figonzal.evaluatool.utilidades.logInfo
import cl.figonzal.evaluatool.utilidades.showHelperDialog
import com.google.android.material.textfield.TextInputEditText

class VelocidadFragmentE7M4 : Fragment(), EvaluaInterface {

    companion object {
        private const val DESVIACION = 68.21
        private const val MEDIA = 257.12

        fun newInstance(): VelocidadFragmentE7M4 {
            return VelocidadFragmentE7M4()
        }
    }

    private var binding: FragmentVelocidadE7M4Binding? = null
    private val perc = listOf(
            100 to 99,
            121 to 97,
            135 to 95,
            150 to 90,
            160 to 85,
            170 to 80,
            180 to 75,
            190 to 70,
            210 to 65,
            220 to 60,
            240 to 55,
            250 to 50,
            260 to 40,
            280 to 30,
            300 to 20,
            340 to 10,
            500 to 5
    )

    //TAREA 1
    private lateinit var etSegundosT1: TextInputEditText
    private var segundosT1 = 0

    //SUBTOTALES
    private lateinit var tvSubTotalT1: TextView
    private var totalPdT1 = 0.0

    //TOTALES
    private lateinit var tvPdTotal: TextView
    private lateinit var tvPdCorregido: TextView
    private lateinit var tvPercentil: TextView
    private lateinit var tvNivel: TextView
    private lateinit var tvDesviacionCalculada: TextView
    private lateinit var progressBar: ProgressBar

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        binding = FragmentVelocidadE7M4Binding.inflate(inflater, container, false)

        initResources(binding!!)
        return binding!!.root
    }

    private fun initResources(binding: FragmentVelocidadE7M4Binding) {

        with(binding, {
            cardViewConstantes.tvMediaValue.text = MEDIA.toString()
            cardViewConstantes.tvDesviacionValue.text = DESVIACION.toString()

            //SUBTOTALES
            tvSubTotalT1 = tvPdSubtotalT1
            this@VelocidadFragmentE7M4.etSegundosT1 = etSegundosT1

            //TOTALES
            this@VelocidadFragmentE7M4.tvPdTotal = tvPdTotalValue
            tvPdCorregido = cardViewFinal.tvPdTotalCorregidoValue
            tvPercentil = cardViewFinal.tvPercentilValue
            tvNivel = cardViewFinal.tvNivelObtenidoValue
            tvDesviacionCalculada = cardViewFinal.tvDesviacionCalculadaValue

            progressBar = cardViewFinal.progressBar
            progressBar.max = perc[0].second

            cardViewFinal.ivHelpPdCorregido.setOnClickListener {

                requireActivity().logInfo(R.string.DIALOGO_AYUDA_MSG_ABIERTO)
                requireActivity().showHelperDialog(requireFragmentManager())
            }

            Utils.configurarTextoBaremo(requireFragmentManager(), tablaBaremo.tvBaremo, perc, getString(R.string.TOOLBAR_VELOCIDAD))
        }).run {
            textWatcherTarea1()
        }
    }

    private fun textWatcherTarea1() {

        with(etSegundosT1) {
            addTextChangedListener(object : TextWatcher {

                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                    totalPdT1 = 0.0
                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

                override fun afterTextChanged(s: Editable) {

                    when {
                        s.isEmpty() -> segundosT1 = 0
                        s.isNotEmpty() -> segundosT1 = text.toString().toInt()
                    }
                    totalPdT1 = calculateTask(null, tvSubTotalT1, context.getString(R.string.TAREA_1), segundosT1, null, null)
                    calculateResult()
                }
            })
        }
    }

    override fun calculateTask(nTarea: Int?, tvSubTotal: TextView, tarea: String, aprobadas: Int?, omitidas: Int?, reprobadas: Int?): Double {
        tvSubTotal.text = String.format(getString(R.string.SEG_FORMAT), tarea, aprobadas)
        return aprobadas!!.toDouble()
    }

    override fun calculateResult() {

        with(totalPdT1, {
            tvPdTotal.text = String.format(getString(R.string.SEG_SIMPLE_FORMAT), this)

            val pdCorregido = correctPD(perc, this)
            tvPdCorregido.text = String.format(getString(R.string.SEG_SIMPLE_FORMAT), pdCorregido)

            tvDesviacionCalculada.text = Utils.calcularDesviacion(MEDIA, DESVIACION, pdCorregido, true).toString()

            with(calculatePercentile(pdCorregido), {
                tvPercentil.text = this.toString()

                when {
                    Build.VERSION.SDK_INT >= Build.VERSION_CODES.N -> progressBar.setProgress(this, true)
                    else -> progressBar.progress = this
                }
                tvNivel.text = Utils.calcularNivel(this)
            })
        })
    }

    override fun calculatePercentile(pdTotal: Double): Int {
        when {
            pdTotal < perc[0].first -> return perc[0].second
            pdTotal > perc[perc.size - 1].first -> return perc[perc.size - 1].second
            else -> perc.forEach { item ->
                if (pdTotal <= item.first) return item.second
            }
        }
        //Percentil no encontrado
        requireActivity().logInfo(R.string.TAG_PERCENTIL_CALCULADO, R.string.PERCENTIL_NULO)
        return -1
    }

    override fun correctPD(perc: List<Pair<Int, Int>>, pdActual: Double): Double {
        when {
            pdActual < perc[0].first -> return perc[0].first.toDouble()
            pdActual > perc[perc.size - 1].first -> return perc[perc.size - 1].first.toDouble()
            else -> perc.forEach { item ->
                if (pdActual <= item.first) return item.first.toDouble()
            }
        }
        requireActivity().logInfo(R.string.TAG_PD_CORREGIDO, R.string.PD_NULO)
        return (-1).toDouble()
    }

    override fun onDestroy() {
        super.onDestroy()
        binding = null
    }
}